src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -98,13 +99,18 @@ private ServiceCallSite CreateCallSite(Type serviceType, CallSiteChain callSiteC
                 return _stackGuard.RunOnEmptyStack((type, chain) => CreateCallSite(type, chain), serviceType, callSiteChain);
             }
 
-            callSiteChain.CheckCircularDependency(serviceType);
+            var callsiteLock = _callSiteLocks.GetOrAdd(serviceType, _ => new object());
 
-            ServiceCallSite callSite = TryCreateExact(serviceType, callSiteChain) ??
-                                       TryCreateOpenGeneric(serviceType, callSiteChain) ??
-                                       TryCreateEnumerable(serviceType, callSiteChain);
+            lock (callsiteLock)
```

Comment:
Can you add a comment that describes what resources this lock protects?  If it's protecting `_callSiteCache` does that still need to be a `ConcurrentDictionary`?

Reply:
I'll add the comment. It does need to be thread safe because it's still modified concurrently. We're just locking the resolution process for a single service type at a time:

Consider the following:

C -> D -> A
E -> D -> A

Resolving C and E in parallel means that they will be modifying the concurrent dictionary to add the entry for C and E in parallel, but the resolution of D and A is synchronized to make sure C and E both reference the same instance of the callsite.